File: problems/132_pattern/solution.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        if not nums: return False
        mn = [nums[0]]
        for n in nums[1:]:
            mn.append(min(mn[-1], n))
        st = []
        for mn, n in zip(mn[::-1], nums[::-1]):
            if mn < n:
                while st and st[-1] <= mn:
                    st.pop()
                if st and st[-1] < n and st[-1] > mn: return True
                st.append(n)
        return False

File: problems/132_pattern/solution.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        vector<int> mns;
        
        int run = INT_MAX;
        for (int n: nums) {
            run = min(run, n);
            mns.push_back(run);
        }
        
        int mn;
        stack<int> st;
        for (int i = nums.size()-1; i >= 0; --i) {
            run = nums[i], mn = mns[i];
            if (run > mn) {
                while (!st.empty() && st.top() <= mn) st.pop();
                if (!st.empty() && st.top() < run) return true;
                st.push(run);
            }
        }
        return false;
    }
};